doc/PTBatcherGUI.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PTBATCHERGUI ""1"""
.TH PTBATCHERGUI "1" "2010-12-26" """Version: 2010.5.0""" "HUGIN"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
PTBatcherGUI \- Hugin batch processor
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBPTBatcherGUI\fR [options] [<project1 <output prefix>> <project2 <output prefix>> <project3...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
A \s-1GUI\s0 Batch Processor for managing the stitching of a queue of hugin .pto projects.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-h|\-\-help\fR" 4
.IX Item "-h|--help"
Show usage message
.IP "\fB\-b|\-\-batch\fR" 4
.IX Item "-b|--batch"
Run batch immediately
.IP "\fB\-p|\-\-parallel\fR" 4
.IX Item "-p|--parallel"
Run batch projects in parallel
.IP "\fB\-d|\-\-delete\fR" 4
.IX Item "-d|--delete"
Delete *.pto files after stitching
.IP "\fB\-o|\-\-overwrite\fR" 4
.IX Item "-o|--overwrite"
Overwrite previous files without asking
.IP "\fB\-s|\-\-shutdown\fR" 4
.IX Item "-s|--shutdown"
Shutdown computer after batch is complete
.IP "\fB\-v|\-\-verbose\fR" 4
.IX Item "-v|--verbose"
Show verbose output when processing projects
.SH "AUTHORS"
.IX Header "AUTHORS"
Written by Pablo d'Angelo, Marko Kuder and others.


NEWS
Newsworthy changes in 0.7 beta 5
--------------------------------
* much improved vignetting and exposure correction,
  http://hugin.sf.net/tech
* ability to create HDR images during stitching
  (still extremly memory hungry, will be improved
   in the next version).
* fixed bad HFOV/VFOV limits of some projections
* OSX build should not crash all the time
* Added "Apply template" function, to apply the
  settings of a previously saved project to
  the current images.
* enabled "load lens data" button in dialog asking
  for focal length/HFOV.

Newsworthy changes in 0.7 beta 4
-------------------------
* update panotools to latest version
* Albers equal area conic and Miller cylindrical projection
* Support for straight line control points
* Fixed running of Autopano-SIFT integration on windows
* (linux) work around bugs in wxWidgets (fixes double deletion
  of control points when pressing the Del key)
* Force even values for width for 360 degree equirectangular
  panoramas.

Newsworthy changes in 0.7 beta 3
-------------------------
* Assistant panel, enables 3 click stitching (well, almost ;-)
* Center and rotate panorama by mouse clicks in preview window.
  Also added a numerical transform dialog
* automatic straighten
* support for Lambert and Lambert Azimuthal projections
* separate crop tab replaces buggy crop dialog
* crop is saved in lens .ini files
* Rotated display of images in control point selector, depending
  on the roll value
* allow images with and without alpha channels in the same project
* option to open final panorama in image editor after stitching
* present output of command line utils in nice window (unix)
* many small bugfixes
* removed image center dialog


Newsworthy changes in 0.6.1
---------------------------
* Autopano-SIFT integration was broken in 0.6 on Unix and OSX.
* Support for displaying HDR files within hugin, using log or gamma mapping
* .hdr file support (tif HDR files have been always useable)
* reduced memory usage, better image caching strategy in GUI.
* fixed crashes and hangs when using the preview auto update and avoid
  unnessecary remapping of images.
* vignetting correction estimation speed up.
* nona and nona_gui can be used to do template stitching, by specifying
  images on the command line (or dnd shortcut under windows)
* distribute fulla.exe with windows package (it was accidentally
  not included in the 0.6 package)
* universal build for Intel Macs, fixed hang after enblend or autopano-sift
  call
* projection can be changed in preview.
* removed all references to malicious panotools.info website
* various smaller bugfixes

Newsworthy changes in 0.6
-------------------------
* updated vigra to version 1.4. All sourcecode of hugin is now
  released under approved open source licences. This means there
  is no problem with including hugin in debian and ubuntu anymore.
* Copy ICC profiles from input images to output
* Added vignetting correction and brighness correction
  during stitching.
  Vignetting can be corrected by addition or division.
  The correction can be done with a flatfield image or an
  radial polynomial
* Functionality to estimate the vignetting polynomial
  directly from the overlapping images. This is an experimental feature.
* Multithreaded remapping. hugin and nona fully utilize multiple
  CPUs or multi core CPUs during preview generation and stitching
* support for dcraw 16 bit .ppm files
* Disabled PTOptimizer progress display on OSX. This should fix
  the crashes some people have experienced.
* New command line tool (fulla), for distortion, TCA and vignetting correction.
  Can use PTLens database (old format only).
* reorganized GUI layout.
* Czech translation added

Newsworthy changes in 0.5
-------------------------
* Added Catalan, Chinese and Hungarian translation
* Support TIFF compressions: NONE, DEFLATE, LZW
* option to create cropped tiff files for usage with enblend 2.4
* Mac OSX: fixed bug during calling PTOptimizer executable
* Windows only:
    hugin is shipped with the pano12.dll build with MSVC.
    This might lead to crashes when using PTStitcher.exe
    In that case replace pano12.dll with pano12_for_usage_with ptstitcher.dll

Newsworthy changes in 0.5 rc2
-----------------------------
* Removed all non-opensource sourcecode from CVS
* Fixed various crashes
* Japanese, Russian, Brazilian and Dutch translation
* Proper default arguments for Autopano-SIFT on Windows
  (users of hugin 0.5 r1 or less need to restore the default
   config)

Newsworthy changes in 0.5 rc1
-----------------------------
* Options on the menu and toolbar that are irrelevant
  are now disabled.
* enblend can now be called from nona even if it has 
  spaces in the path.

Newsworthy changes in 0.5 beta7
-------------------------------
* Updated translations
* Fixes to the build system on linux and windows

Newsworthy changes in 0.5 beta6
-------------------------------
* All files are now in source tarball
* French translation now complete

Newsworthy changes in 0.5 beta5
-------------------------------

* Support for modifying the size of the right/left panes in many tabs.
* All controls can now be reached on smaller displays as many areas 
  of the interface are now scrollable.
* All image control buttons in the preview tab can now be accessed, 
  allowing support for far more images.

* OSX build now supports unicode.

* Enblend arguments specified in preferences are now passed on correctly.
* Multiple tiff file outputs are no longer deleted incorrectly.
* Simultaneous scrolling in control points editor tabs respects borders properly.


src/hugin1/ptbatcher/PTBatcherGUI.cpp
// -*- c-basic-offset: 4 -*-

/** @file PTBatcherGUI.cpp
 *
 *  @brief Batch processor for Hugin with GUI
 *
 *  @author Marko Kuder <marko.kuder@gmail.com>
 *
 *  $Id: PTBatcherGUI.cpp 3322 2008-08-16 5:00:07Z mkuder $
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this software; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include "PTBatcherGUI.h"
#ifdef __WXMSW__
#include "wx/cshelp.h"
#endif

// make wxwindows use this class as the main application
IMPLEMENT_APP(PTBatcherGUI)

BEGIN_EVENT_TABLE(PTBatcherGUI, wxApp)
	EVT_LIST_ITEM_ACTIVATED(XRCID("project_listbox"),PTBatcherGUI::OnItemActivated)
	EVT_KEY_DOWN(PTBatcherGUI::OnKeyDown)
END_EVENT_TABLE()

bool PTBatcherGUI::OnInit()
{
	// Required to access the preferences of hugin
    SetAppName(wxT("hugin"));

#if defined __WXMSW__
	int localeID = wxConfigBase::Get()->Read(wxT("language"), (long) wxLANGUAGE_DEFAULT);
	m_locale.Init(localeID);
    // initialize help provider
    wxHelpControllerHelpProvider* provider = new wxHelpControllerHelpProvider;
    wxHelpProvider::Set(provider);

#else
    m_locale.Init(wxLANGUAGE_DEFAULT);
#endif

    // setup the environment for the different operating systems
#if defined __WXMSW__
    wxString huginExeDir = getExePath(argv[0]);

    wxString huginRoot;
    wxFileName::SplitPath(huginExeDir, &huginRoot, NULL, NULL);
	m_xrcPrefix = wxString(huginRoot + wxT("/share/hugin/xrc/"));
	
	// locale setup
    m_locale.AddCatalogLookupPathPrefix(huginRoot + wxT("/share/locale"));

#elif defined __WXMAC__ && defined MAC_SELF_CONTAINED_BUNDLE
    {
        wxString exec_path = MacGetPathToBundledResourceFile(CFSTR("xrc"));
        if(exec_path != wxT(""))
        {
			m_xrcPrefix = exec_path + wxT("/");
        }
		else
			wxMessageBox(_("xrc directory not found in bundle"), _("Fatal Error"));

        wxString thePath = MacGetPathToBundledResourceFile(CFSTR("locale"));
        if(thePath != wxT(""))
            m_locale.AddCatalogLookupPathPrefix(thePath);
        else {
            wxMessageBox(_("Translations not found in bundle"), _("Fatal Error"));
            return false;
        }
    }
#else
    // add the locale directory specified during configure
	m_xrcPrefix = wxT(INSTALL_XRC_DIR);
    m_locale.AddCatalogLookupPathPrefix(wxT(INSTALL_LOCALE_DIR));
#endif

    // set the name of locale recource to look for
    m_locale.AddCatalog(wxT("hugin"));

	const wxString name = wxString::Format(_T("PTBatcherGUI-%s"), wxGetUserId().c_str());
	m_checker = new wxSingleInstanceChecker(name+wxT(".lock"),wxFileName::GetTempDir());
	bool IsFirstInstance=(!m_checker->IsAnotherRunning());

	if(IsFirstInstance)
	{
		if ( ! wxFile::Exists(m_xrcPrefix + wxT("/batch_frame.xrc")) ) {
			wxMessageBox(_("xrc directory not found, hugin needs to be properly installed\nTried Path:") + m_xrcPrefix , _("Fatal Error"));
			return false;
		}
		// initialize image handlers
		wxInitAllImageHandlers();

		// Initialize all the XRC handlers.
		wxXmlResource::Get()->InitAllHandlers();
		wxXmlResource::Get()->AddHandler(new ProjectListBoxXmlHandler());
		// load XRC files
		wxXmlResource::Get()->Load(m_xrcPrefix + wxT("batch_frame.xrc"));
		wxXmlResource::Get()->Load(m_xrcPrefix + wxT("batch_toolbar.xrc"));
		wxXmlResource::Get()->Load(m_xrcPrefix + wxT("batch_menu.xrc"));
	};

	// parse arguments
    static const wxCmdLineEntryDesc cmdLineDesc[] =
    {
        //On wxWidgets 2.9, wide characters don't work here.
        //On previous versions, the wxT macro is required for unicode builds.
#if wxCHECK_VERSION(2,9,0)
      { wxCMD_LINE_SWITCH, "h", "help", "show this help message",
        wxCMD_LINE_VAL_NONE, wxCMD_LINE_OPTION_HELP },
      { wxCMD_LINE_SWITCH, "b", "batch",  "run batch immediately" },
      { wxCMD_LINE_SWITCH, "p", "parallel",  "run batch projects in parallel" },
      { wxCMD_LINE_SWITCH, "d", "delete",  "delete *.pto files after stitching" },
      { wxCMD_LINE_SWITCH, "o", "overwrite",  "overwrite previous files without asking" },
      { wxCMD_LINE_SWITCH, "s", "shutdown",  "shutdown computer after batch is complete" },
      { wxCMD_LINE_SWITCH, "v", "verbose",  "show verbose output when processing projects" },
      { wxCMD_LINE_SWITCH, "a", "assistant", "run the assistant on the given projects" },
      { wxCMD_LINE_PARAM,  NULL, NULL, _("stitch_project.pto [output prefix]|assistant_project.pto"),
        wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL + wxCMD_LINE_PARAM_MULTIPLE },
      { wxCMD_LINE_NONE }
#else
      { wxCMD_LINE_SWITCH, wxT("h"), wxT("help"), wxT("show this help message"),
        wxCMD_LINE_VAL_NONE, wxCMD_LINE_OPTION_HELP },
      { wxCMD_LINE_SWITCH, wxT("b"), wxT("batch"),  wxT("run batch immediately") },
	  { wxCMD_LINE_SWITCH, wxT("p"), wxT("parallel"),  wxT("run batch projects in parallel") },
	  { wxCMD_LINE_SWITCH, wxT("d"), wxT("delete"),  wxT("delete *.pto files after stitching") },
	  { wxCMD_LINE_SWITCH, wxT("o"), wxT("overwrite"),  wxT("overwrite previous files without asking") },
	  { wxCMD_LINE_SWITCH, wxT("s"), wxT("shutdown"),  wxT("shutdown computer after batch is complete") },
	  { wxCMD_LINE_SWITCH, wxT("v"), wxT("verbose"),  wxT("show verbose output when processing projects") },
      { wxCMD_LINE_SWITCH, wxT("a"), wxT("assistant"), wxT("run the assistant on the given projects") },
      { wxCMD_LINE_PARAM,  NULL, NULL, _("stitch_project.pto [output prefix]|assistant_project.pto"),
        wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL + wxCMD_LINE_PARAM_MULTIPLE },
      { wxCMD_LINE_NONE }
#endif
    };
    wxCmdLineParser parser(cmdLineDesc, argc, argv);

    switch ( parser.Parse() ) {
		case -1: // -h or --help was given, and help displayed so exit
			return false;
			break;
		case 0:  // all is well
			break;
		default:
			wxLogError(_("Syntax error in parameters detected, aborting."));
			return false;
			break;
    }

	wxClient client;
	wxConnectionBase *conn;
	wxString servername;
#ifdef __WINDOWS__
	servername=name;
#else
	servername=wxFileName::GetTempDir()+wxFileName::GetPathSeparator()+name+wxT(".ipc");
#endif
	if(IsFirstInstance)
	{
		m_frame = new BatchFrame(&m_locale,m_xrcPrefix);
		m_frame->RestoreSize();
#ifdef __WXMSW__
        provider->SetHelpController(&m_frame->GetHelpController());
        m_frame->GetHelpController().Initialize(m_xrcPrefix+wxT("data/hugin_help_en_EN.chm"));
#endif
		SetTopWindow(m_frame);
		m_frame->Show(true);
		m_server = new BatchIPCServer();
		if (!m_server->Create(servername))
		{
			delete m_server;
			m_server = NULL;
		};
	}
	else
	{
		conn=client.MakeConnection(wxEmptyString, servername, IPC_START);
		if(!conn)
			return false;
	};
	//m_frame->SetLocaleAndXRC(&m_locale,m_xrcPrefix);
	//projectsRunning=0;
	unsigned int count = 0;
    if(parser.Found(wxT("a")))
    {
        //added assistant files
	    while(parser.GetParamCount()>count)
        {
	    	wxString param = parser.GetParam(count);
		    count++;
			wxFileName name(param);
    	    name.MakeAbsolute();
            if(name.FileExists())
            {
                //only add existing pto files
                if(name.GetExt().CmpNoCase(wxT("pto"))==0)
                {
                    if(IsFirstInstance)
                    {
                        m_frame->AddToList(name.GetFullPath(),Project::DETECTING);
                    }
	    		    else
                    {
			    	    conn->Request(wxT("D ")+name.GetFullPath());
                    };
                };
            };
        };
    }
    else
    {
    	bool projectSpecified = false;
	    //we collect all parameters - all project files <and their output prefixes>
	    while(parser.GetParamCount()>count)
    	{
	    	wxString param = parser.GetParam(count);
		    count++;
		    if(!projectSpecified)	//next parameter must be new script file
		    {
			    wxFileName name(param);
    			name.MakeAbsolute();
	    		if(IsFirstInstance)
		    		m_frame->AddToList(name.GetFullPath());
			    else
				    conn->Request(wxT("A ")+name.GetFullPath());
    			projectSpecified = true;
	    	}
		    else	//parameter could be previous project's output prefix
    		{
	    		wxFileName fn(param);
		    	fn.MakeAbsolute();
			    if(!fn.HasExt())	//if there is no extension we have a prefix
    			{
	    			if(IsFirstInstance)
		    			m_frame->ChangePrefix(-1,fn.GetFullPath());
			    	else
				    	conn->Request(wxT("P ")+fn.GetFullPath());
    				projectSpecified = false;
	    		}
		    	else
			    {
				    wxString ext = fn.GetExt();
    				//we may still have a prefix, but with added image extension
	    			if (ext.CmpNoCase(wxT("jpg")) == 0 || ext.CmpNoCase(wxT("jpeg")) == 0|| 
		    			ext.CmpNoCase(wxT("tif")) == 0|| ext.CmpNoCase(wxT("tiff")) == 0 || 
			    		ext.CmpNoCase(wxT("png")) == 0 || ext.CmpNoCase(wxT("exr")) == 0 ||
				    	ext.CmpNoCase(wxT("pnm")) == 0 || ext.CmpNoCase(wxT("hdr")) == 0) 
				    {
					    //extension will be removed before stitch, so there is no need to do it now
    					if(IsFirstInstance)
	    					m_frame->ChangePrefix(-1,fn.GetFullPath());
		    			else
			    			conn->Request(wxT("P ")+fn.GetFullPath());
				    	projectSpecified = false;
    				}
	    			else //if parameter has a different extension we presume it is a new script file
    				{
    					//we add the new project
	    				if(IsFirstInstance)
		    				m_frame->AddToList(fn.GetFullPath());
			    		else
				    		conn->Request(wxT("A ")+fn.GetFullPath());
					    projectSpecified = true;
    				}
			    } //else of if(!fn.HasExt())
		    }
	    }
    }
    
#ifdef __WXMAC__
    m_macFileNameToOpenOnStart = wxT("");
    wxYield();
    if(m_macFileNameToOpenOnStart != wxT(""))
    {
        wxFileName fn(m_macFileNameToOpenOnStart);
        m_frame->AddToList(fn.GetFullPath());
    }
#endif

	if(IsFirstInstance)
	{
		wxConfigBase *config=wxConfigBase::Get();
		if (parser.Found(wxT("d")))
			config->Write(wxT("/BatchFrame/DeleteCheck"), 1l);
		if (parser.Found(wxT("p")))
			config->Write(wxT("/BatchFrame/ParallelCheck"), 1l);
		if (parser.Found(wxT("s")))
			config->Write(wxT("/BatchFrame/ShutdownCheck"), 1l);
		if (parser.Found(wxT("o")))
			config->Write(wxT("/BatchFrame/OverwriteCheck"), 1l);
		if (parser.Found(wxT("v")))
			config->Write(wxT("/BatchFrame/VerboseCheck"), 1l);
		config->Flush();
	}
	else
	{
		if (parser.Found(wxT("d")))
			conn->Request(wxT("SetDeleteCheck"));
		if (parser.Found(wxT("p")))
			conn->Request(wxT("SetParallelCheck"));
		if (parser.Found(wxT("s")))
			conn->Request(wxT("SetShutdownCheck"));
		if (parser.Found(wxT("o")))
			conn->Request(wxT("SetOverwriteCheck"));
		if (parser.Found(wxT("v")))
			conn->Request(wxT("SetVerboseCheck"));
		conn->Request(wxT("BringWindowToTop"));
		if(parser.Found(wxT("b")))
			conn->Request(wxT("RunBatch"));
		conn->Disconnect();
		delete conn;
		delete m_checker;
		return false;
	};
	m_frame->SetCheckboxes();
	m_frame->PropagateDefaults();
	if (parser.Found(wxT("b")) ) {
		m_frame->RunBatch();
    }
	return true;
}

int PTBatcherGUI::OnExit()
{
	delete m_checker;
	delete m_server;
	return 0;
}

void PTBatcherGUI::OnItemActivated(wxListEvent &event)
{
	wxCommandEvent dummy;
	m_frame->OnButtonOpenWithHugin(dummy);
}

void PTBatcherGUI::OnKeyDown(wxKeyEvent &event)
{
	wxCommandEvent dummy;
	switch(event.GetKeyCode())
	{
	case WXK_DELETE:	
		m_frame->OnButtonRemoveFromList(dummy);
		break;
	case WXK_INSERT:
		m_frame->OnButtonAddToList(dummy);
		break;
	case WXK_ESCAPE:
		m_frame->OnButtonCancel(dummy);
		break;
	default:
		event.Skip();
		break;
	}
	
}

#ifdef __WXMAC__
// wx calls this method when the app gets "Open file" AppleEvent
void PTBatcherGUI::MacOpenFile(const wxString &fileName)
{
    if(!m_frame) 
        m_macFileNameToOpenOnStart = fileName;
    else
    {
        wxFileName fn(fileName);
        m_frame->AddToList(fn.GetFullPath());
    }
}
#endif

wxChar* BatchIPCConnection::OnRequest(const wxString& topic, const wxString& item, int *size, wxIPCFormat format)
{
	BatchFrame *MyBatchFrame=wxGetApp().GetFrame();
	if(item.Left(1)==wxT("A"))
    {
		MyBatchFrame->AddToList(item.Mid(2));
        return wxT("");
    };
	if(item.Left(1)==wxT("D"))
    {
        MyBatchFrame->AddToList(item.Mid(2),Project::DETECTING);
        return wxT("");
    };
	if(item.Left(1)==wxT("P"))
    {
		MyBatchFrame->ChangePrefix(-1,item.Mid(2));
        return wxT("");
    };
	wxCommandEvent event;
	event.SetInt(1);
	if(item==wxT("SetDeleteCheck"))
		if(!MyBatchFrame->GetCheckDelete())
		{
			MyBatchFrame->OnCheckDelete(event);
			MyBatchFrame->SetCheckboxes();
		};
	if(item==wxT("SetParallelCheck"))
		if(!MyBatchFrame->GetCheckParallel())
		{
			MyBatchFrame->OnCheckParallel(event);
			MyBatchFrame->SetCheckboxes();
		};
	if(item==wxT("SetShutdownCheck"))
		if(!MyBatchFrame->GetCheckShutdown())
		{
			MyBatchFrame->OnCheckShutdown(event);
			MyBatchFrame->SetCheckboxes();
		};
	if(item==wxT("SetOverwriteCheck"))
		if(!MyBatchFrame->GetCheckOverwrite())
		{
			MyBatchFrame->OnCheckOverwrite(event);
			MyBatchFrame->SetCheckboxes();
		};
	if(item==wxT("SetVerboseCheck"))
		if(!MyBatchFrame->GetCheckVerbose())
		{
			MyBatchFrame->OnCheckVerbose(event);
			MyBatchFrame->SetCheckboxes();
		};
	if(item==wxT("BringWindowToTop"))
		MyBatchFrame->RequestUserAttention();
	if(item==wxT("RunBatch"))
	{
		wxCommandEvent myEvent(wxEVT_COMMAND_TOOL_CLICKED ,XRCID("tool_start"));
		MyBatchFrame->GetEventHandler()->AddPendingEvent(myEvent);
	};
	return wxT("");
};

wxConnectionBase* BatchIPCServer::OnAcceptConnection (const wxString& topic)
{
	if(topic==IPC_START)
		return new BatchIPCConnection;
	return NULL;
};
